[ICE]: parallel: None in compiler/rustc_type_ir/src/ty_kind.rs

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
//@compile-flags: --edition=2024  -Zthreads=20
#![feature(never_patterns)]
trait A {
    fn g() -> B;
}

trait B {
    fn bar(&self, x: &A);
}

fn main() {}

```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.96.0-nightly (d933cf483 2026-03-04)
binary: rustc
commit-hash: d933cf483edf1605142ac6899ff32536c0ad8b22
commit-date: 2026-03-04
host: x86_64-unknown-linux-gnu
release: 1.96.0-nightly
LLVM version: 22.1.0
```

### Error output

```
<output>
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
warning: the feature `never_patterns` is incomplete and may not be safe to use and/or cause compiler crashes
 --> code.rs:2:12
  |
2 | #![feature(never_patterns)]
  |            ^^^^^^^^^^^^^^
  |
  = note: see issue #118155 <https://github.com/rust-lang/rust/issues/118155> for more information
  = note: `#[warn(incomplete_features)]` on by default


thread 'rustc' (1168854) panicked at /rustc-dev/d933cf483edf1605142ac6899ff32536c0ad8b22/compiler/rustc_type_ir/src/ty_kind.rs:787:60:
called `Option::unwrap()` on a `None` value
stack backtrace:
   0:     0x7f46efd5ef7b - <<std[69d51b4813ba037]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[51d4adbef590f61]::fmt::Display>::fmt
   1:     0x7f46f04229c8 - core[51d4adbef590f61]::fmt::write
   2:     0x7f46efd75f86 - <std[69d51b4813ba037]::sys::stdio::unix::Stderr as std[69d51b4813ba037]::io::Write>::write_fmt
   3:     0x7f46efd35058 - std[69d51b4813ba037]::panicking::default_hook::{closure#0}
   4:     0x7f46efd523e3 - std[69d51b4813ba037]::panicking::default_hook
   5:     0x7f46eed3d71c - std[69d51b4813ba037]::panicking::update_hook::<alloc[18203fec9a0cd4ec]::boxed::Box<rustc_driver_impl[9032f3c5cdcc149c]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7f46efd526c2 - std[69d51b4813ba037]::panicking::panic_with_hook
   7:     0x7f46efd3514a - std[69d51b4813ba037]::panicking::panic_handler::{closure#0}
   8:     0x7f46efd2be49 - std[69d51b4813ba037]::sys::backtrace::__rust_end_short_backtrace::<std[69d51b4813ba037]::panicking::panic_handler::{closure#0}, !>
   9:     0x7f46efd36b7d - __rustc[cb16bc19fc993bd7]::rust_begin_unwind
  10:     0x7f46ec91895c - core[51d4adbef590f61]::panicking::panic_fmt
  11:     0x7f46ec3fea02 - core[51d4adbef590f61]::panicking::panic
  12:     0x7f46ecfefc69 - core[51d4adbef590f61]::option::unwrap_failed
  13:     0x7f46f0d2d4b3 - rustc_trait_selection[a0d03008c221415d]::traits::dyn_compatibility::virtual_call_violations_for_method
  14:     0x7f46f160bc22 - rustc_trait_selection[a0d03008c221415d]::traits::dyn_compatibility::dyn_compatibility_violations_for_assoc_item
  15:     0x7f46f160b6d8 - <core[51d4adbef590f61]::iter::adapters::flatten::FlatMap<core[51d4adbef590f61]::iter::adapters::map::Map<core[51d4adbef590f61]::iter::adapters::map::Map<core[51d4adbef590f61]::slice::iter::Iter<(core[51d4adbef590f61]::option::Option<rustc_span[a0277d8dc90b615]::symbol::Symbol>, rustc_middle[7e63d2df2dd45427]::ty::assoc::AssocItem)>, <rustc_data_structures[889238d3fa4068cd]::sorted_map::index_map::SortedIndexMultiMap<u32, core[51d4adbef590f61]::option::Option<rustc_span[a0277d8dc90b615]::symbol::Symbol>, rustc_middle[7e63d2df2dd45427]::ty::assoc::AssocItem>>::iter::{closure#0}>, <rustc_middle[7e63d2df2dd45427]::ty::assoc::AssocItems>::in_definition_order::{closure#0}>, alloc[18203fec9a0cd4ec]::vec::Vec<rustc_middle[7e63d2df2dd45427]::traits::DynCompatibilityViolation>, rustc_trait_selection[a0d03008c221415d]::traits::dyn_compatibility::dyn_compatibility_violations_for_trait::{closure#0}::{closure#0}> as core[51d4adbef590f61]::iter::traits::iterator::Iterator>::next
  16:     0x7f46f16096e9 - <core[51d4adbef590f61]::iter::adapters::GenericShunt<core[51d4adbef590f61]::iter::adapters::map::Map<core[51d4adbef590f61]::iter::adapters::flatten::FlatMap<core[51d4adbef590f61]::iter::sources::from_fn::FromFn<rustc_type_ir[505b55bac6a24d17]::elaborate::supertrait_def_ids<rustc_middle[7e63d2df2dd45427]::ty::context::TyCtxt>::{closure#0}>, alloc[18203fec9a0cd4ec]::vec::Vec<rustc_middle[7e63d2df2dd45427]::traits::DynCompatibilityViolation>, rustc_trait_selection[a0d03008c221415d]::traits::dyn_compatibility::dyn_compatibility_violations::{closure#0}>, core[51d4adbef590f61]::result::Result<rustc_middle[7e63d2df2dd45427]::traits::DynCompatibilityViolation, !>::Ok>, core[51d4adbef590f61]::result::Result<core[51d4adbef590f61]::convert::Infallible, !>> as core[51d4adbef590f61]::iter::traits::iterator::Iterator>::next
  17:     0x7f46f1608abc - rustc_trait_selection[a0d03008c221415d]::traits::dyn_compatibility::dyn_compatibility_violations
  18:     0x7f46f16089ac - rustc_query_impl[540cdfee4f8a4756]::query_impl::dyn_compatibility_violations::invoke_provider_fn::__rust_begin_short_backtrace
  19:     0x7f46f06ecfb7 - rustc_query_impl[540cdfee4f8a4756]::execution::try_execute_query::<rustc_middle[7e63d2df2dd45427]::query::caches::DefIdCache<rustc_middle[7e63d2df2dd45427]::query::erase::ErasedData<[u8; 16usize]>>, false>
  20:     0x7f46f0ddec27 - rustc_query_impl[540cdfee4f8a4756]::query_impl::dyn_compatibility_violations::execute_query_non_incr::__rust_end_short_backtrace
  21:     0x7f46f0ddecc0 - rustc_trait_selection[a0d03008c221415d]::traits::dyn_compatibility::is_dyn_compatible
  22:     0x7f46f0ddec57 - rustc_query_impl[540cdfee4f8a4756]::query_impl::is_dyn_compatible::invoke_provider_fn::__rust_begin_short_backtrace
  23:     0x7f46f0dd854a - rustc_query_impl[540cdfee4f8a4756]::execution::try_execute_query::<rustc_middle[7e63d2df2dd45427]::query::caches::DefIdCache<rustc_middle[7e63d2df2dd45427]::query::erase::ErasedData<[u8; 1usize]>>, false>
  24:     0x7f46f0dd6fc0 - rustc_query_impl[540cdfee4f8a4756]::query_impl::is_dyn_compatible::execute_query_non_incr::__rust_end_short_backtrace
  25:     0x7f46f17e64be - <dyn rustc_hir_analysis[bceba1a8102fadd5]::hir_ty_lowering::HirTyLowerer>::prohibit_or_lint_bare_trait_object_ty
  26:     0x7f46f12476ff - <dyn rustc_hir_analysis[bceba1a8102fadd5]::hir_ty_lowering::HirTyLowerer>::lower_trait_object_ty
  27:     0x7f46f0ab45bc - <dyn rustc_hir_analysis[bceba1a8102fadd5]::hir_ty_lowering::HirTyLowerer>::lower_ty
  28:     0x7f46f0acaeaf - <dyn rustc_hir_analysis[bceba1a8102fadd5]::hir_ty_lowering::HirTyLowerer>::lower_fn_ty
  29:     0x7f46f06f14bc - rustc_hir_analysis[bceba1a8102fadd5]::collect::fn_sig
  30:     0x7f46f06f0c26 - rustc_query_impl[540cdfee4f8a4756]::query_impl::fn_sig::invoke_provider_fn::__rust_begin_short_backtrace
  31:     0x7f46f06f5360 - rustc_query_impl[540cdfee4f8a4756]::execution::try_execute_query::<rustc_middle[7e63d2df2dd45427]::query::caches::DefIdCache<rustc_middle[7e63d2df2dd45427]::query::erase::ErasedData<[u8; 24usize]>>, false>
  32:     0x7f46f06f4853 - rustc_query_impl[540cdfee4f8a4756]::query_impl::fn_sig::execute_query_non_incr::__rust_end_short_backtrace
  33:     0x7f46f0adf473 - rustc_hir_analysis[bceba1a8102fadd5]::check::check::check_item_type
  34:     0x7f46f0addca6 - rustc_hir_analysis[bceba1a8102fadd5]::check::wfcheck::check_well_formed
  35:     0x7f46f0addc87 - rustc_query_impl[540cdfee4f8a4756]::query_impl::check_well_formed::invoke_provider_fn::__rust_begin_short_backtrace
  36:     0x7f46f0add219 - rustc_query_impl[540cdfee4f8a4756]::execution::try_execute_query::<rustc_data_structures[889238d3fa4068cd]::vec_cache::VecCache<rustc_span[a0277d8dc90b615]::def_id::LocalDefId, rustc_middle[7e63d2df2dd45427]::query::erase::ErasedData<[u8; 1usize]>, rustc_middle[7e63d2df2dd45427]::dep_graph::graph::DepNodeIndex>, false>
  37:     0x7f46f0adcf7b - rustc_query_impl[540cdfee4f8a4756]::query_impl::check_well_formed::execute_query_non_incr::__rust_end_short_backtrace
  38:     0x7f46eeea25ca - <rustc_middle[7e63d2df2dd45427]::hir::ModuleItems>::par_opaques::<rustc_hir_analysis[bceba1a8102fadd5]::check::wfcheck::check_type_wf::{closure#5}>::{closure#0}
  39:     0x7f46eef34c1b - <rustc_thread_pool[ad99c7f29618ca7]::job::HeapJob<<rustc_thread_pool[ad99c7f29618ca7]::scope::Scope>::spawn<rustc_data_structures[889238d3fa4068cd]::sync::parallel::par_slice<&rustc_span[a0277d8dc90b615]::def_id::LocalDefId, rustc_data_structures[889238d3fa4068cd]::sync::parallel::try_par_for_each_in<&[rustc_span[a0277d8dc90b615]::def_id::LocalDefId], rustc_span[a0277d8dc90b615]::ErrorGuaranteed, <rustc_middle[7e63d2df2dd45427]::hir::ModuleItems>::par_opaques<rustc_hir_analysis[bceba1a8102fadd5]::check::wfcheck::check_type_wf::{closure#5}>::{closure#0}>::{closure#0}::{closure#0}>::{closure#0}::{closure#0}>::{closure#0}> as rustc_thread_pool[ad99c7f29618ca7]::job::Job>::execute
  40:     0x7f46efa53c9f - <rustc_thread_pool[ad99c7f29618ca7]::registry::WorkerThread>::wait_or_steal_until_cold
  41:     0x7f46efa52832 - <rustc_thread_pool[ad99c7f29618ca7]::registry::ThreadBuilder>::run
  42:     0x7f46eed361dc - std[69d51b4813ba037]::sys::backtrace::__rust_begin_short_backtrace::<<rustc_thread_pool[ad99c7f29618ca7]::ThreadPoolBuilder>::build_scoped<rustc_interface[5b554394f8ce662a]::util::run_in_thread_pool_with_globals<rustc_interface[5b554394f8ce662a]::interface::run_compiler<(), rustc_driver_impl[9032f3c5cdcc149c]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#5}::{closure#0}::{closure#0}, rustc_interface[5b554394f8ce662a]::util::run_in_thread_pool_with_globals<rustc_interface[5b554394f8ce662a]::interface::run_compiler<(), rustc_driver_impl[9032f3c5cdcc149c]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#5}::{closure#0}::{closure#1}, ()>::{closure#0}::{closure#0}::{closure#0}, ()>
  43:     0x7f46eed402e3 - <std[69d51b4813ba037]::thread::lifecycle::spawn_unchecked<<rustc_thread_pool[ad99c7f29618ca7]::ThreadPoolBuilder>::build_scoped<rustc_interface[5b554394f8ce662a]::util::run_in_thread_pool_with_globals<rustc_interface[5b554394f8ce662a]::interface::run_compiler<(), rustc_driver_impl[9032f3c5cdcc149c]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#5}::{closure#0}::{closure#0}, rustc_interface[5b554394f8ce662a]::util::run_in_thread_pool_with_globals<rustc_interface[5b554394f8ce662a]::interface::run_compiler<(), rustc_driver_impl[9032f3c5cdcc149c]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#5}::{closure#0}::{closure#1}, ()>::{closure#0}::{closure#0}::{closure#0}, ()>::{closure#1} as core[51d4adbef590f61]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  44:     0x7f46f165dbec - <std[69d51b4813ba037]::sys::thread::unix::Thread>::new::thread_start
  45:     0x7f46eb05e98b - <unknown>
  46:     0x7f46eb0e2a0c - <unknown>
  47:                0x0 - <unknown>

error: the compiler unexpectedly panicked. This is a bug

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/home/matthias/vcs/github/CRED/rustc-ice-2026-03-04T08_23_51-1168850.txt` to your bug report

note: rustc 1.96.0-nightly (d933cf483 2026-03-04) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=8

query stack during panic:
#0 [dyn_compatibility_violations] determining dyn-compatibility of trait `B`
#1 [is_dyn_compatible] checking if trait `B` is dyn-compatible
#2 [fn_sig] computing function signature of `A::g`
#3 [check_well_formed] checking that `A::g` is well-formed
#4 [check_type_wf] checking that types are well-formed
#5 [analysis] running analysis passes on crate `code`
end of query stack
warning: 1 warning emitted

note: no errors encountered even though delayed bugs were created

note: those delayed bugs will now be shown as internal compiler errors

error: internal compiler error[E0391]: cycle detected when computing function signature of `A::g`
 --> code.rs:4:5
  |
4 |     fn g() -> B;
  |     ^^^^^^^^^^^^
  |
note: ...which requires checking if trait `B` is dyn-compatible...
 --> code.rs:7:1
  |
7 | trait B {
  | ^^^^^^^
note: ...which requires determining dyn-compatibility of trait `B`...
 --> code.rs:7:1
  |
7 | trait B {
  | ^^^^^^^
note: ...which requires computing function signature of `B::bar`...
 --> code.rs:8:5
  |
8 |     fn bar(&self, x: &A);
  |     ^^^^^^^^^^^^^^^^^^^^^
note: ...which requires checking if trait `A` is dyn-compatible...
 --> code.rs:3:1
  |
3 | trait A {
  | ^^^^^^^
note: ...which requires determining dyn-compatibility of trait `A`...
 --> code.rs:3:1
  |
3 | trait A {
  | ^^^^^^^
  = note: ...which again requires computing function signature of `A::g`, completing the cycle
note: cycle used when checking that `A::g` is well-formed
 --> code.rs:4:5
  |
4 |     fn g() -> B;
  |     ^^^^^^^^^^^^
  = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information
note: delayed at /rustc-dev/d933cf483edf1605142ac6899ff32536c0ad8b22/compiler/rustc_query_impl/src/job.rs:482:16
         0: <rustc_errors::DiagCtxtInner>::emit_diagnostic
         1: <rustc_errors::DiagCtxtHandle>::emit_diagnostic
         2: <rustc_span::ErrorGuaranteed as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
         3: <rustc_errors::diagnostic::Diag>::delay_as_bug
         4: rustc_query_impl::execution::mk_cycle::<rustc_middle::query::caches::DefIdCache<rustc_middle::query::erase::ErasedData<[u8; 24]>>>
         5: rustc_query_impl::execution::try_execute_query::<rustc_middle::query::caches::DefIdCache<rustc_middle::query::erase::ErasedData<[u8; 24]>>, false>
         6: rustc_query_impl::query_impl::fn_sig::execute_query_non_incr::__rust_end_short_backtrace
         7: rustc_trait_selection::traits::dyn_compatibility::virtual_call_violations_for_method
         8: rustc_trait_selection::traits::dyn_compatibility::dyn_compatibility_violations_for_assoc_item
         9: <core::iter::adapters::flatten::FlatMap<core::iter::adapters::map::Map<core::iter::adapters::map::Map<core::slice::iter::Iter<(core::option::Option<rustc_span::symbol::Symbol>, rustc_middle::ty::assoc::AssocItem)>, <rustc_data_structures::sorted_map::index_map::SortedIndexMultiMap<u32, core::option::Option<rustc_span::symbol::Symbol>, rustc_middle::ty::assoc::AssocItem>>::iter::{closure#0}>, <rustc_middle::ty::assoc::AssocItems>::in_definition_order::{closure#0}>, alloc::vec::Vec<rustc_middle::traits::DynCompatibilityViolation>, rustc_trait_selection::traits::dyn_compatibility::dyn_compatibility_violations_for_trait::{closure#0}::{closure#0}> as core::iter::traits::iterator::Iterator>::next
        10: <core::iter::adapters::GenericShunt<core::iter::adapters::map::Map<core::iter::adapters::flatten::FlatMap<core::iter::sources::from_fn::FromFn<rustc_type_ir::elaborate::supertrait_def_ids<rustc_middle::ty::context::TyCtxt>::{closure#0}>, alloc::vec::Vec<rustc_middle::traits::DynCompatibilityViolation>, rustc_trait_selection::traits::dyn_compatibility::dyn_compatibility_violations::{closure#0}>, core::result::Result<rustc_middle::traits::DynCompatibilityViolation, !>::Ok>, core::result::Result<core::convert::Infallible, !>> as core::iter::traits::iterator::Iterator>::next
        11: rustc_trait_selection::traits::dyn_compatibility::dyn_compatibility_violations
        12: rustc_query_impl::query_impl::dyn_compatibility_violations::invoke_provider_fn::__rust_begin_short_backtrace
        13: rustc_query_impl::execution::try_execute_query::<rustc_middle::query::caches::DefIdCache<rustc_middle::query::erase::ErasedData<[u8; 16]>>, false>
        14: rustc_query_impl::query_impl::dyn_compatibility_violations::execute_query_non_incr::__rust_end_short_backtrace
        15: rustc_trait_selection::traits::dyn_compatibility::is_dyn_compatible
        16: rustc_query_impl::query_impl::is_dyn_compatible::invoke_provider_fn::__rust_begin_short_backtrace
        17: rustc_query_impl::execution::try_execute_query::<rustc_middle::query::caches::DefIdCache<rustc_middle::query::erase::ErasedData<[u8; 1]>>, false>
        18: rustc_query_impl::query_impl::is_dyn_compatible::execute_query_non_incr::__rust_end_short_backtrace
        19: <dyn rustc_hir_analysis::hir_ty_lowering::HirTyLowerer>::prohibit_or_lint_bare_trait_object_ty
        20: <dyn rustc_hir_analysis::hir_ty_lowering::HirTyLowerer>::lower_trait_object_ty
        21: <dyn rustc_hir_analysis::hir_ty_lowering::HirTyLowerer>::lower_ty
        22: <dyn rustc_hir_analysis::hir_ty_lowering::HirTyLowerer>::lower_fn_ty
        23: rustc_hir_analysis::collect::fn_sig
        24: rustc_query_impl::query_impl::fn_sig::invoke_provider_fn::__rust_begin_short_backtrace
        25: rustc_query_impl::execution::try_execute_query::<rustc_middle::query::caches::DefIdCache<rustc_middle::query::erase::ErasedData<[u8; 24]>>, false>
        26: rustc_query_impl::query_impl::fn_sig::execute_query_non_incr::__rust_end_short_backtrace
        27: rustc_hir_analysis::check::check::check_item_type
        28: rustc_hir_analysis::check::wfcheck::check_well_formed
        29: rustc_query_impl::query_impl::check_well_formed::invoke_provider_fn::__rust_begin_short_backtrace
        30: rustc_query_impl::execution::try_execute_query::<rustc_data_structures::vec_cache::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::ErasedData<[u8; 1]>, rustc_middle::dep_graph::graph::DepNodeIndex>, false>
        31: rustc_query_impl::query_impl::check_well_formed::execute_query_non_incr::__rust_end_short_backtrace
        32: <rustc_middle::hir::ModuleItems>::par_opaques::<rustc_hir_analysis::check::wfcheck::check_type_wf::{closure#5}>::{closure#0}
        33: <rustc_thread_pool::job::HeapJob<<rustc_thread_pool::scope::Scope>::spawn<rustc_data_structures::sync::parallel::par_slice<&rustc_span::def_id::LocalDefId, rustc_data_structures::sync::parallel::try_par_for_each_in<&[rustc_span::def_id::LocalDefId], rustc_span::ErrorGuaranteed, <rustc_middle::hir::ModuleItems>::par_opaques<rustc_hir_analysis::check::wfcheck::check_type_wf::{closure#5}>::{closure#0}>::{closure#0}::{closure#0}>::{closure#0}::{closure#0}>::{closure#0}> as rustc_thread_pool::job::Job>::execute
        34: <rustc_thread_pool::registry::WorkerThread>::wait_or_steal_until_cold
        35: <rustc_thread_pool::registry::ThreadBuilder>::run
        36: std::sys::backtrace::__rust_begin_short_backtrace::<<rustc_thread_pool::ThreadPoolBuilder>::build_scoped<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#5}::{closure#0}::{closure#0}, rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#5}::{closure#0}::{closure#1}, ()>::{closure#0}::{closure#0}::{closure#0}, ()>
        37: <std::thread::lifecycle::spawn_unchecked<<rustc_thread_pool::ThreadPoolBuilder>::build_scoped<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#5}::{closure#0}::{closure#0}, rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#5}::{closure#0}::{closure#1}, ()>::{closure#0}::{closure#0}::{closure#0}, ()>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
        38: <std::sys::thread::unix::Thread>::new::thread_start
        39: <unknown>
        40: <unknown>

 --> code.rs:4:5
  |
4 |     fn g() -> B;
  |     ^^^^^^^^^^^^

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/home/matthias/vcs/github/CRED/rustc-ice-2026-03-04T08_23_51-1168850.txt` to your bug report

note: rustc 1.96.0-nightly (d933cf483 2026-03-04) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=8

query stack during panic:
end of query stack
```

</p>
</details>


<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_middle/src/query/plumbing.rs
+++ b/compiler/rustc_middle/src/query/plumbing.rs
@@ -136,8 +136,12 @@ pub struct QueryVTable<'tcx, C: QueryCache> {
     /// For `no_hash` queries, this function pointer is None.
     pub hash_value_fn: Option<fn(&mut StableHashingContext<'_>, &C::Value) -> Fingerprint>,
 
-    pub value_from_cycle_error:
-        fn(tcx: TyCtxt<'tcx>, cycle_error: CycleError, guar: ErrorGuaranteed) -> C::Value,
+    pub value_from_cycle_error: fn(
+        tcx: TyCtxt<'tcx>,
+        key: C::Key,
+        cycle_error: CycleError,
+        guar: ErrorGuaranteed,
+    ) -> C::Value,
     pub format_value: fn(&C::Value) -> String,
 
     /// Formats a human-readable description of this query and its key, as

--- a/compiler/rustc_query_impl/src/execution.rs
+++ b/compiler/rustc_query_impl/src/execution.rs
@@ -125,17 +125,18 @@ where
 fn mk_cycle<'tcx, C: QueryCache>(
     query: &'tcx QueryVTable<'tcx, C>,
     tcx: TyCtxt<'tcx>,
+    key: C::Key,
     cycle_error: CycleError,
 ) -> C::Value {
     let error = report_cycle(tcx.sess, &cycle_error);
     match query.cycle_error_handling {
         CycleErrorHandling::Error => {
             let guar = error.emit();
-            (query.value_from_cycle_error)(tcx, cycle_error, guar)
+            (query.value_from_cycle_error)(tcx, key, cycle_error, guar)
         }
         CycleErrorHandling::DelayBug => {
             let guar = error.delay_as_bug();
-            (query.value_from_cycle_error)(tcx, cycle_error, guar)
+            (query.value_from_cycle_error)(tcx, key, cycle_error, guar)
         }
     }
 }
@@ -219,6 +220,7 @@ where
 fn cycle_error<'tcx, C: QueryCache>(
     query: &'tcx QueryVTable<'tcx, C>,
     tcx: TyCtxt<'tcx>,
+    key: C::Key,
     try_execute: QueryJobId,
     span: Span,
 ) -> (C::Value, Option<DepNodeIndex>) {
@@ -229,7 +231,7 @@ fn cycle_error<'tcx, C: QueryCache>(
         .expect("failed to collect active queries");
 
     let error = find_cycle_in_stack(try_execute, job_map, &current_query_job(), span);
-    (mk_cycle(query, tcx, error.lift()), None)
+    (mk_cycle(query, tcx, key, error.lift()), None)
 }
 
 #[inline(always)]
@@ -274,7 +276,7 @@ fn wait_for_query<'tcx, C: QueryCache>(
 
             (v, Some(index))
         }
-        Err(cycle) => (mk_cycle(query, tcx, cycle.lift()), None),
+        Err(cycle) => (mk_cycle(query, tcx, key, cycle.lift()), None),
     }
 }
 
@@ -337,7 +339,7 @@ fn try_execute_query<'tcx, C: QueryCache, const INCR: bool>(
 
                         // If we are single-threaded we know that we have cycle error,
                         // so we just return the error.
-                        cycle_error(query, tcx, id, span)
+                        cycle_error(query, tcx, key, id, span)
                     }
                 }
                 ActiveKeyStatus::Poisoned => FatalError.raise(),

--- a/compiler/rustc_query_impl/src/from_cycle_error.rs
+++ b/compiler/rustc_query_impl/src/from_cycle_error.rs
@@ -15,34 +15,34 @@ use rustc_middle::query::erase::erase_val;
 use rustc_middle::ty::layout::{LayoutError, TyAndLayout};
 use rustc_middle::ty::{self, Ty, TyCtxt};
 use rustc_middle::{bug, span_bug};
-use rustc_span::def_id::LocalDefId;
+use rustc_span::def_id::{DefId, LocalDefId};
 use rustc_span::{ErrorGuaranteed, Span};
 
 use crate::job::report_cycle;
 
 pub(crate) fn specialize_query_vtables<'tcx>(vtables: &mut QueryVTables<'tcx>) {
     vtables.type_of.value_from_cycle_error =
-        |tcx, _, guar| erase_val(ty::EarlyBinder::bind(Ty::new_error(tcx, guar)));
+        |tcx, _, _, guar| erase_val(ty::EarlyBinder::bind(Ty::new_error(tcx, guar)));
 
     vtables.type_of_opaque_hir_typeck.value_from_cycle_error =
-        |tcx, _, guar| erase_val(ty::EarlyBinder::bind(Ty::new_error(tcx, guar)));
+        |tcx, _, _, guar| erase_val(ty::EarlyBinder::bind(Ty::new_error(tcx, guar)));
 
     vtables.erase_and_anonymize_regions_ty.value_from_cycle_error =
-        |tcx, _, guar| erase_val(Ty::new_error(tcx, guar));
+        |tcx, _, _, guar| erase_val(Ty::new_error(tcx, guar));
 
-    vtables.fn_sig.value_from_cycle_error = |tcx, cycle, guar| erase_val(fn_sig(tcx, cycle, guar));
+    vtables.fn_sig.value_from_cycle_error = |tcx, key, _, guar| erase_val(fn_sig(tcx, key, guar));
 
     vtables.check_representability.value_from_cycle_error =
-        |tcx, cycle, guar| check_representability(tcx, cycle, guar);
+        |tcx, _, cycle, guar| check_representability(tcx, cycle, guar);
 
     vtables.check_representability_adt_ty.value_from_cycle_error =
-        |tcx, cycle, guar| check_representability(tcx, cycle, guar);
+        |tcx, _, cycle, guar| check_representability(tcx, cycle, guar);
 
     vtables.variances_of.value_from_cycle_error =
-        |tcx, cycle, guar| erase_val(variances_of(tcx, cycle, guar));
+        |tcx, _, cycle, guar| erase_val(variances_of(tcx, cycle, guar));
 
     vtables.layout_of.value_from_cycle_error =
-        |tcx, cycle, guar| erase_val(layout_of(tcx, cycle, guar));
+        |tcx, _, cycle, guar| erase_val(layout_of(tcx, cycle, guar));
 }
 
 pub(crate) fn default<'tcx>(tcx: TyCtxt<'tcx>, cycle_error: CycleError, query_name: &str) -> ! {
@@ -57,15 +57,12 @@ pub(crate) fn default<'tcx>(tcx: TyCtxt<'tcx>, cycle_error: CycleError, query_na
 
 fn fn_sig<'tcx>(
     tcx: TyCtxt<'tcx>,
-    cycle_error: CycleError,
+    def_id: DefId,
     guar: ErrorGuaranteed,
 ) -> ty::EarlyBinder<'tcx, ty::PolyFnSig<'tcx>> {
     let err = Ty::new_error(tcx, guar);
 
-    let arity = if let Some(info) = cycle_error.cycle.get(0)
-        && info.frame.dep_kind == DepKind::fn_sig
-        && let Some(def_id) = info.frame.def_id
-        && let Some(node) = tcx.hir_get_if_local(def_id)
+    let arity = if let Some(node) = tcx.hir_get_if_local(def_id)
         && let Some(sig) = node.fn_sig()
     {
         sig.decl.inputs.len()

--- a/compiler/rustc_query_impl/src/plumbing.rs
+++ b/compiler/rustc_query_impl/src/plumbing.rs
@@ -487,7 +487,7 @@ macro_rules! define_queries {
                     #[cfg(not($cache_on_disk))]
                     is_loadable_from_disk_fn: |_tcx, _key, _index| false,
 
-                    value_from_cycle_error: |tcx, cycle, _| {
+                    value_from_cycle_error: |tcx, _, cycle, _| {
                         $crate::from_cycle_error::default(tcx, cycle, stringify!($name))
                     },
 

--- a/tests/ui/parallel-rustc/fn-sig-cycle-ice-153391.rs
+++ b/tests/ui/parallel-rustc/fn-sig-cycle-ice-153391.rs
@@ -1,0 +1,18 @@
+// Regression test for #153391.
+//
+//@ edition:2024
+//@ compile-flags: -Z threads=16
+//@ compare-output-by-lines
+//@ ignore-test (#142063)
+
+trait A {
+    fn g() -> B;
+    //~^ ERROR expected a type, found a trait
+}
+
+trait B {
+    fn bar(&self, x: &A);
+    //~^ ERROR expected a type, found a trait
+}
+
+fn main() {}

--- a/tests/ui/parallel-rustc/fn-sig-cycle-ice-153391.stderr
+++ b/tests/ui/parallel-rustc/fn-sig-cycle-ice-153391.stderr
@@ -1,0 +1,31 @@
+error[E0782]: expected a type, found a trait
+  --> $DIR/fn-sig-cycle-ice-153391.rs:8:15
+   |
+LL |     fn g() -> B;
+   |               ^
+   |
+help: `B` is dyn-incompatible, use `impl B` to return an opaque type, as long as you return a single underlying type
+   |
+LL |     fn g() -> impl B;
+   |               ++++
+
+error[E0782]: expected a type, found a trait
+  --> $DIR/fn-sig-cycle-ice-153391.rs:13:23
+   |
+LL |     fn bar(&self, x: &A);
+   |                       ^
+   |
+   = note: `A` is dyn-incompatible, otherwise a trait object could be used
+help: use a new generic type parameter, constrained by `A`
+   |
+LL -     fn bar(&self, x: &A);
+LL +     fn bar<T: A>(&self, x: &T);
+   |
+help: you can also use an opaque type, but users won't be able to specify the type parameter when calling the `fn`, having to rely exclusively on type inference
+   |
+LL |     fn bar(&self, x: &impl A);
+   |                       ++++
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0782`.

